src/acquisition/covidcast/covidcast_row.py

```diff
@@ -0,0 +1,270 @@
+from dataclasses import asdict, dataclass, field, fields
+from datetime import date
+from typing import Any, ClassVar, Dict, Iterable, List, Optional
+
+import pandas as pd
+from delphi_utils import Nans
+
+from delphi.epidata.acquisition.covidcast.csv_importer import CsvImporter
+from delphi.epidata.server.utils.dates import day_to_time_value, time_value_to_day
+from delphi.epidata.server.endpoints.covidcast_utils.model import PANDAS_DTYPES
+
+
+@dataclass
+class CovidcastRow:
+    """A container for the values of a single covidcast database row.
+
+    Used for:
+    - inserting rows into the database
+    - creating test rows with default fields for testing
+    - converting from and to formats (dict, csv, df, kwargs)
+    - creating consistent views, with consistent data types (dict, csv, df)
+
+    The rows are specified in 'v4_schema.sql'. The datatypes are made to match database. When writing to Pandas, the dtypes match the JIT model.py schema.
+    """
+
+    # Arguments.
+    source: str = "src"
+    signal: str = "sig"
+    time_type: str = "day"
+    geo_type: str = "county"
+    time_value: int = 20200202
+    geo_value: str = "01234"
+    value: float = 10.0
+    stderr: float = 10.0
+    sample_size: float = 10.0
+    missing_value: int = Nans.NOT_MISSING.value
+    missing_stderr: int = Nans.NOT_MISSING.value
+    missing_sample_size: int = Nans.NOT_MISSING.value
+    issue: Optional[int] = 20200202
+    lag: Optional[int] = 0
+    id: Optional[int] = None
+    direction: Optional[int] = None
+    direction_updated_timestamp: int = 0
+    value_updated_timestamp: int = 20200202
+
+    # Classvars.
+    _api_row_ignore_fields: ClassVar = ["id", "direction_updated_timestamp", "value_updated_timestamp"]
+    _api_row_compatibility_ignore_fields: ClassVar = ["id", "direction_updated_timestamp", "value_updated_timestamp", "source"]
+    _db_row_ignore_fields: ClassVar = []
+    _pandas_dtypes: ClassVar = PANDAS_DTYPES
+
+    def __post_init__(self):
+        # Convert time values to ints by default.
+        self.time_value = day_to_time_value(self.time_value) if isinstance(self.time_value, date) else self.time_value
+        self.issue = day_to_time_value(self.issue) if isinstance(self.issue, date) else self.issue
+        self.value_updated_timestamp = day_to_time_value(self.value_updated_timestamp) if isinstance(self.value_updated_timestamp, date) else self.value_updated_timestamp
+
+    def _sanity_check_fields(self, extra_checks: bool = True):
+        if self.issue and self.issue < self.time_value:
+            self.issue = self.time_value
+
+        if self.issue:
+            self.lag = (time_value_to_day(self.issue) - time_value_to_day(self.time_value)).days
+        else:
+            self.lag = None
+
+        # This sanity checking is already done in CsvImporter, but it's here so the testing class gets it too.
```

Comment:
is there a matching note in CsvImporter, so we remember to update it here too if we ever need to change it there?

Reply:
Alternately, why not have CsvImporter call this code?

Reply:
Will make a note.

It would be nice to integrate CsvImporter with this code, agreed. Should probably make a separate PR though.